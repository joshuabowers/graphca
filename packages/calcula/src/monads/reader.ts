export type Reader<E, T> = (env: E) => T

export const unit = <E, T>(t: T): Reader<E, T> => _ => t

export const bind = <E, A>(r: Reader<E, A>) => 
  <B>(fn: (a: A) => Reader<E, B>): Reader<E, B> =>
    (e: E) => fn(r(e))(e)

export const ask = <E>(env: E) => env

export const local = <E, T>(fn: (env: E) => E) =>
  (c: Reader<E, T>) =>
    (e: E) => c(fn(e))
