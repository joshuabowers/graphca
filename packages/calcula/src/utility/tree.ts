import { Writer } from "../monads/writer"

export enum Clades {
  primitive = 'Primitive',  // Constants: 10, 4 - 5i, true
  unary = 'Unary',          // Functions and Ops: not, abs, sin
  binary = 'Binary',        // Functions and Ops: and, log, polygamma
  variadic = 'Variadic'     // Variables: x, OhGodWhy
}

export enum Genera {
  arithmetic = 'Arithmetic',
  connective = 'Connective',
  inequalities = 'Inequalities',
  trigonometric = 'Trigonometric',
  arcus = 'Arcus',
  hyperbolic = 'Hyperbolic',
  areaHyperbolic = 'AreaHyperbolic',
  combinatorics = 'Combinatorics'
}

export enum Species {
  // Primitives
  real = 'Real',
  complex = 'Complex',
  boolean = 'Boolean',
  nil = 'Nil',
  nan = 'NaN',

  // Arithmetic
  add = 'Addition',
  multiply = 'Multiplication',
  raise = 'Exponentiation',

  // Functions
  abs = 'Absolute',

  // :: Inequalities
  equals = 'Equality',
  notEquals = 'Inequality',
  lessThan = 'LessThan',
  greaterThan = 'GreaterThan',
  lessThanEquals = 'LessThanEquals',
  greaterThanEquals = 'GreaterThanEquals',

  // :: Connectives
  not = 'Complement',
  and = 'Conjunction',
  or = 'Disjunction',
  xor = 'ExclusiveDisjunction',
  implies = 'Implication',
  nand = 'AlternativeDenial',
  nor = 'JointDenial',
  xnor = 'Biconditional',
  converse = 'ConverseImplication',

  // :: Trigonometric
  sin = 'Sine',
  cos = 'Cosine',
  tan = 'Tangent',
  csc = 'Cosecant',
  sec = 'Secant',
  cot = 'Cotangent',

  // Variables
  variable = 'Variable'
}

export type TreeNode = {
  readonly clade: Clades,
  readonly genus?: Genera,
  readonly species: Species
}

export type DerivedNode<T extends ((...args: any[]) => any)> = 
  ReturnType<T> extends Writer<infer U> ? U : never

export type TreeNodeGuardFn<T extends TreeNode> = 
  (value: Writer<TreeNode>) => value is Writer<T>

// values should be Writer<TreeNode>, narrowed to Writer<T>
export const isClade = <T extends TreeNode>(clade: Clades) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.clade === clade

export const isGenus = <T extends TreeNode>(genus: Genera) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.genus === genus

export const isSpecies = <T extends TreeNode>(species: Species) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.species === species

export const notAny = (...args: Species[]) => {
  const exclude = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && !exclude.has(t.value.species)
}

export const any = (...args: Species[]) => {
  const include = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && include.has(t.value.species)
}
