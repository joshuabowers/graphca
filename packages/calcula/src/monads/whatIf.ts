import { multi, method, Multi } from "@arrows/multimethod"
import { Species, Clades, Genera } from "../utility/tree"
import { 
  Real, Complex, Boolean,
  real, complex, boolean
} from "../primitives"

export interface Writer<Value, Operation> {
  value: Value,
  log: Operation[]
}

export type WriterFn<V, O, N = V> = (result: V) => Writer<N, O>

export const unit = <V, O>(result: V): Writer<V, O> => ({value: result, log: []})

export const bind = <V, O, N = V>(
  writer: Writer<V, O>, 
  transform: WriterFn<V, O, N>
): Writer<N, O> => {
  const {value: current, log} = writer
  const {value: result, log: updates} = transform(current)
  return {value: result, log: [...log, ...updates]}
}

export const isWriter = <V, O>(obj: unknown): obj is Writer<V, O> =>
  typeof obj === 'object' && ('value' in (obj ?? {})) && ('log' in (obj ?? {}))

export type TreeNode = {
  readonly species: Species,
  readonly genus?: Genera,
  readonly clade: Clades
}

export type Classify<
  N extends TreeNode, 
  S extends Species, 
  G extends Genera|undefined
> =
  N & Readonly<{species: S, genus?: G}>

export type UnaryNode = TreeNode & {
  readonly clade: Clades.unary,
  readonly argument: TreeNode
}

export type Unary<S extends Species, G extends Genera|undefined = undefined> = 
  Classify<UnaryNode, S, G>

export type BinaryNode = TreeNode & {
  readonly clade: Clades.binary,
  readonly left: TreeNode,
  readonly right: TreeNode
}

export type Binary<S extends Species, G extends Genera> = Classify<BinaryNode, S, G>

export type MultiaryNode = TreeNode & {
  readonly clade: Clades.multiary,
  readonly operands: TreeNode[]
}

export type Multiary<S extends Species, G extends Genera> = Classify<MultiaryNode, S, G>

export enum Alter {
  insert = 'i',
  remove = 'r',
  update = 'u'
}

export type Path = number[]
export type Particle = string | Particle[]

export type Delta = [Alter, Path, Particle]

export type Operation = [string, ...Delta[]]

// const o: Operation = [
//   'referenced variable',
//   [Alter.update, [0, 1, 0, 2], ['(', 'x', ')']],
//   [Alter.insert, [0, 0], ['(', 'y', '+', '3', ')']]
// ]

// const [action, ...deltas] = o

// type Overload<Params extends any[], Result extends TreeNode> = 
//   ((...params: Params) => Writer<Result, Operation>)

type UnaryOverload<Argument extends TreeNode, Result extends TreeNode> =
  (argument: Argument|Writer<Argument, Operation>) => Writer<Result, Operation>

type UnaryFn<T extends UnaryNode> = Multi
  & UnaryOverload<TreeNode, T>

type UnaryCreateFn<T extends UnaryNode> = (argument: TreeNode) => [T, string]

type UnaryPredicate<Argument extends TreeNode> = (argument: Argument) => boolean
type UnaryCorrespondingFn<Argument extends TreeNode, Result extends TreeNode> =
  (argument: Argument) => [Result, string]

// This, implicitly, would not work in a way that behaves similarly to
// logFunctional: mArgument's logs would be inserted into the amalgamated
// log first, followed by whatever is generated by fn. However, logFunctional
// would have a preamble that contextualizes mArguments logs.
//
// logFunctional structure: (e.g. for abs(x))
// 1. species identified ("abs(x)")
// 2. processing argument ("abs([x])")
// 3. arguments logs ("x")
// 4. processed argument ("abs({x})")
// 5. result logs ("abs(x)")
//
// 2-4 repeated (with slight alterations to denote what is affected) for
// binary/multiary contexts.
const unaryBind = <
  T extends UnaryNode,
  Argument extends TreeNode, 
  Result extends TreeNode
>(
  create: UnaryCreateFn<T>, species: Species,
  fn: UnaryCorrespondingFn<Argument, Result>
) =>
  (mArgument: Writer<Argument, Operation>) =>
    bind(unit(create(mArgument.value)[0]), initial => 
      bind(mArgument, argument => {
        const [result, action] = fn(argument)
        const o: Operation = [action, [Alter.insert, [0], ['']]]
        return ({value: result, log: [o]}) // <- logFunctional?
      })
    )

const ensureWriter = <
  T extends UnaryNode,
  Argument extends TreeNode, 
  Result extends TreeNode
>(
  fn: ReturnType<typeof unaryBind<T, Argument, Result>>
) => (input: Argument|Writer<Argument, Operation>) =>
  fn(isWriter(input) ? input : unit(input))

const when = <Argument extends TreeNode, Result extends TreeNode = TreeNode>(
  predicate: UnaryPredicate<Argument>,
  fn: UnaryCorrespondingFn<Argument, Result>
) => 
  <T extends UnaryNode>(create: UnaryCreateFn<T>, species: Species) => 
    method(predicate, ensureWriter(unaryBind(create, species, fn)))

type WhenFn = ReturnType<typeof when>

const unary = <T extends UnaryNode>(
  name: string, species: Species, genus?: Genera
) => {
  const create: UnaryCreateFn<T> = (argument) => [
    ({ clade: Clades.unary, species, genus, argument }) as T,
    `created ${species.toLocaleLowerCase()}`
  ]

  // Primitive Handlers
  return () => {
    // Extensions
    return (...extensions: WhenFn[]) => {
      const fn: UnaryFn<T> = multi(
        ...extensions.map(e => e(create, species))
      )
      return [fn]
    }
  }
}

export type Absolute = Unary<Species.abs>

const isTreeNode = (value: unknown): value is TreeNode =>
  typeof value === 'object' && value! && 'species' in value

const isAbsolute = (value: unknown): value is Absolute =>
  isTreeNode(value) && value.species === Species.abs

export const [abs] = unary<Absolute>('abs', Species.abs)(
  // Primitive handlers
)(
  // Extensions
  when<Absolute>(
    isAbsolute,
    a => [a.argument, 'absolute redundancy']
  )
)
