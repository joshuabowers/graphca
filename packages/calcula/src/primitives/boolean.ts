import { Species, isSpecies } from '../utility/tree'
import { Boolean, primitive } from '../closures/primitive'
import { isBoolean as isSystemBoolean } from '../utility/valuePredicates'
export { Boolean }

export const boolean = primitive<boolean, {value: boolean}, Boolean>(
  isSystemBoolean,
  value => ({value}),
  Species.boolean
)(
  create => r => [create(r.value !== 0), 'cast to boolean'],
  create => c => [create(c.a !== 0 || c.b !== 0), 'cast to boolean'],
  _create => b => [b, '']
)()

export const isBoolean = isSpecies<Boolean>(Species.boolean)
