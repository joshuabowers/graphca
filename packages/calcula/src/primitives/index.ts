import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
