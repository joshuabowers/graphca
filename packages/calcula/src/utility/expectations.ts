import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
