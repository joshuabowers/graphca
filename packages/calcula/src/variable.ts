import { Writer, unit } from './monads/writer'
import { TreeNode, Clades, Species, isSpecies } from './utility/tree'
import { isNil, nil } from './primitives/nil'
import { rule, process, resolve } from './utility/rule'

export type Variable = TreeNode & {
  readonly clade: Clades.variadic,
  readonly species: Species.variable,
  readonly name: string,
  readonly value: Writer<TreeNode>
}

export const isVariable = isSpecies<Variable>(Species.variable)

export const variable = (
  name: string, value: Writer<TreeNode> = nil
): Writer<Variable> => 
  ({
    value: {
      clade: Clades.variadic,
      species: Species.variable,
      name, value
    },
    log: [{
      input: process`${name}`, 
      rewrite: isNil(value) ? resolve`${name}` : process`${value}`,
      action: isNil(value) 
        ? 'referenced variable' : 'referenced variable with value'
    }] 
  })

export type Scope = Map<string, Writer<Variable>>

export const isScope = (v: unknown): v is Scope =>
  typeof v === 'object' && v !== null
  && ['get', 'set', 'has', 'delete'].every(m => m in v)

type Entries = Iterable<readonly [string, Writer<Variable>]>

export const scope = (entries: Entries = []): Scope => 
  new Map<string, Writer<Variable>>(entries)

export const assign = (name: string, value: Writer<TreeNode>, scope?: Scope) => {
  const v = variable(name, value)
  if(scope){
    if(isNil(value)){ scope.delete(name) }
    else { scope.set(name, v) }
  }
  return v
}
