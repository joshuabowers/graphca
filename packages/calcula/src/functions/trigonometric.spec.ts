import { unit } from '../monads/writer'
import { expectCloseTo, expectWriterTreeNode } from '../utility/expectations'
import { Clades, Genera, Species } from '../utility/tree'
import { real, complex } from '../primitives'
import { variable } from '../variable'
import {
  cos, sin, tan, sec, csc, cot,
  $cos, $sin, $tan, $sec, $csc, $cot
} from './trigonometric'

describe('$cos', () => {
  it('generates a Cosine for a TreeNode input', () => {
    expect($cos(unit(variable('x').value))[0]).toEqual({
      clade: Clades.unary, genus: Genera.trigonometric, species: Species.cos,
      expression: unit(variable('x').value)
    })
  })
})

describe('cos', () => {
  it('returns the appropriate real value when given real', () => {
    const v = Math.cos(0.5).toString()
    expectWriterTreeNode(
      cos(real(0.5)),
      real(Math.cos(0.5))
    )(
      ['cos(0.5)', 'identified cosine'],
      ['cos(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['cos(0.5|)', 'processed argument'],
      ['cos(0.5)', 'real cosine'],
      [v, 'created real']
    )
  })

  it('returns a complex value when given complex', () => {
    expectCloseTo(cos(complex([1, 2])), complex([2.032723007019, -3.051897799151]), 10)
  })

  it('returns an expression when given non-constant', () => {
    expectWriterTreeNode(
      cos(variable('x')),
      $cos(unit(variable('x').value))[0]
    )(
      ['cos(x)', 'identified cosine'],
      ['cos(|x)', 'processing argument'],
      ['x', 'referenced variable'],
      ['cos(x|)', 'processed argument'],
      ['cos(x)', 'created cosine']
    )
  })
})

describe('$sin', () => {
  it('generates a Sine for a TreeNode input', () => {
    expect($sin(unit(variable('x').value))[0]).toEqual({
      clade: Clades.unary, genus: Genera.trigonometric, species: Species.sin,
      expression: unit(variable('x').value)
    })
  })
})

describe('sin', () => {
  it('returns the appropriate real value when given real', () => {
    const v = Math.sin(0.5).toString()
    expectWriterTreeNode(
      sin(real(0.5)),
      real(Math.sin(0.5))
    )(
      ['sin(0.5)', 'identified sine'],
      ['sin(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['sin(0.5|)', 'processed argument'],
      ['sin(0.5)', 'real sine'],
      [v, 'created real']
    )
  })

  it('returns a complex value when given complex', () => {
    expectCloseTo(sin(complex([1, 2])), complex([3.16577851321616, 1.95960104142160]), 10)
  })

  it('returns a expression when given non-constant', () => {
    expectWriterTreeNode(
      sin(variable('x')),
      $sin(unit(variable('x').value))[0]
    )(
      ['sin(x)', 'identified sine'],
      ['sin(|x)', 'processing argument'],
      ['x', 'referenced variable'],
      ['sin(x|)', 'processed argument'],
      ['sin(x)', 'created sine']
    )
  })
})

describe('$tan', () => {
  it('generates a Tangent for a TreeNode input', () => {
    expect($tan(unit(variable('x').value))[0]).toEqual({
      clade: Clades.unary, genus: Genera.trigonometric, species: Species.tan,
      expression: unit(variable('x').value)
    })
  })
})

describe('tan', () => {
  it('returns the appropriate real value when given real', () => {
    const v = Math.tan(0.5).toString()
    expectWriterTreeNode(
      tan(real(0.5)),
      real(Math.tan(0.5))
    )(
      ['tan(0.5)', 'identified tangent'],
      ['tan(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['tan(0.5|)', 'processed argument'],
      ['tan(0.5)', 'real tangent'],
      [v, 'created real']
    )
  })

  it('returns a complex value when given complex', () => {
    expectCloseTo(tan(complex([1, 2])), complex([0.033812826079, 1.014793616146]), 10)
  })

  it('returns a expression when given non-constant', () => {
    expectWriterTreeNode(
      tan(variable('x')),
      $tan(unit(variable('x').value))[0]
    )(
      ['tan(x)', 'identified tangent'],
      ['tan(|x)', 'processing argument'],
      ['x', 'referenced variable'],
      ['tan(x|)', 'processed argument'],
      ['tan(x)', 'created tangent']
    )
  })
})

describe('$sec', () => {
  it('generates a Secant for a TreeNode input', () => {
    expect($sec(unit(variable('x').value))[0]).toEqual({
      clade: Clades.unary, genus: Genera.trigonometric, species: Species.sec,
      expression: unit(variable('x').value)
    })
  })
})


describe('sec', () => {
  it('returns the appropriate real value when given real', () => {
    const v = (1 / Math.cos(0.5)).toString()
    const w = Math.cos(0.5).toString()
    expectWriterTreeNode(
      sec(real(0.5)),
      real(1 / Math.cos(0.5))
    )(
      ['sec(0.5)', 'identified secant'],
      ['sec(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['sec(0.5|)', 'processed argument'],
      ['sec(0.5)', 'real secant'],
      ['(cos(0.5)^-1)', 'identified exponentiation'],
      ['(|cos(0.5)^-1)', 'processing left operand'],
      ['cos(0.5)', 'identified cosine'],
      ['cos(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['cos(0.5|)', 'processed argument'],
      ['cos(0.5)', 'real cosine'],
      [w, 'created real'],
      [`(${w}^|-1)`, 'processed left operand; processing right operand'],
      ['-1', 'created real'],
      [`(${w}^-1|)`, 'processed right operand'],
      [`(${w}^-1)`, 'real exponentiation'],
      [v, 'created real']
    )
  })

  it('returns a complex value when given complex', () => {
    expectCloseTo(sec(complex([1, 2])), complex([0.15117629826, 0.22697367539]), 10)
  })

  it('returns a expression when given non-constant', () => {
    expectWriterTreeNode(
      sec(variable('x')),
      $sec(unit(variable('x').value))[0]
    )(
      ['sec(x)', 'identified secant'],
      ['sec(|x)', 'processing argument'],
      ['x', 'referenced variable'],
      ['sec(x|)', 'processed argument'],
      ['sec(x)', 'created secant']
    )
  })
})

describe('$csc', () => {
  it('generates a Cosecant for a TreeNode input', () => {
    expect($csc(unit(variable('x').value))[0]).toEqual({
      clade: Clades.unary, genus: Genera.trigonometric, species: Species.csc,
      expression: unit(variable('x').value)
    })
  })
})


describe('csc', () => {
  it('returns the appropriate real value when given real', () => {
    const v = (1 / Math.sin(0.5)).toString()
    const w = Math.sin(0.5).toString()
    expectWriterTreeNode(
      csc(real(0.5)),
      real(1 / Math.sin(0.5))
    )(
      ['csc(0.5)', 'identified cosecant'],
      ['csc(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['csc(0.5|)', 'processed argument'],
      ['csc(0.5)', 'real cosecant'],
      ['(sin(0.5)^-1)', 'identified exponentiation'],
      ['(|sin(0.5)^-1)', 'processing left operand'],
      ['sin(0.5)', 'identified sine'],
      ['sin(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['sin(0.5|)', 'processed argument'],
      ['sin(0.5)', 'real sine'],
      [w, 'created real'],
      [`(${w}^|-1)`, 'processed left operand; processing right operand'],
      ['-1', 'created real'],
      [`(${w}^-1|)`, 'processed right operand'],
      [`(${w}^-1)`, 'real exponentiation'],
      [v, 'created real']
    )
  })

  it('returns a complex value when given complex', () => {
    expectCloseTo(csc(complex([1, 2])), complex([0.228375065599, -0.141363021612]), 10)
  })

  it('returns a expression when given non-constant', () => {
    expectWriterTreeNode(
      csc(variable('x')),
      $csc(unit(variable('x').value))[0]
    )(
      ['csc(x)', 'identified cosecant'],
      ['csc(|x)', 'processing argument'],
      ['x', 'referenced variable'],
      ['csc(x|)', 'processed argument'],
      ['csc(x)', 'created cosecant']
    )
  })
})

describe('$cot', () => {
  it('generates a Cotangent for a TreeNode input', () => {
    expect($cot(unit(variable('x').value))[0]).toEqual({
      clade: Clades.unary, genus: Genera.trigonometric, species: Species.cot,
      expression: unit(variable('x').value)
    })
  })
})

describe('cot', () => {
  it('returns the appropriate real value when given real', () => {
    const v = (1 / Math.tan(0.5)).toString()
    const w = Math.tan(0.5).toString()
    expectWriterTreeNode(
      cot(real(0.5)),
      real(1 / Math.tan(0.5))
    )(
      ['cot(0.5)', 'identified cotangent'],
      ['cot(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['cot(0.5|)', 'processed argument'],
      ['cot(0.5)', 'real cotangent'],
      ['(tan(0.5)^-1)', 'identified exponentiation'],
      ['(|tan(0.5)^-1)', 'processing left operand'],
      ['tan(0.5)', 'identified tangent'],
      ['tan(|0.5)', 'processing argument'],
      ['0.5', 'created real'],
      ['tan(0.5|)', 'processed argument'],
      ['tan(0.5)', 'real tangent'],
      [w, 'created real'],
      [`(${w}^|-1)`, 'processed left operand; processing right operand'],
      ['-1', 'created real'],
      [`(${w}^-1|)`, 'processed right operand'],
      [`(${w}^-1)`, 'real exponentiation'],
      [v, 'created real']
    )
  })

  it('returns a complex value when given complex', () => {
    expectCloseTo(cot(complex([1, 2])), complex([0.0327977555337, -0.9843292264581]), 10)
  })

  it('returns a expression when given non-constant', () => {
    expectWriterTreeNode(
      cot(variable('x')),
      $cot(unit(variable('x').value))[0]
    )(
      ['cot(x)', 'identified cotangent'],
      ['cot(|x)', 'processing argument'],
      ['x', 'referenced variable'],
      ['cot(x|)', 'processed argument'],
      ['cot(x)', 'created cotangent']
    )
  })
})
