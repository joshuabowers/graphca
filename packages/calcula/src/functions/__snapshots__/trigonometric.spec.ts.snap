// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cos returns an expression when given non-constant 1`] = `
[
  [
    "cos(x)",
    "identified cosine",
  ],
  [
    "cos([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "cos({x})",
    "processed argument",
  ],
  [
    "cos(x)",
    "created cosine",
  ],
]
`;

exports[`cos returns the appropriate real value when given real 1`] = `
[
  [
    "cos(0.5)",
    "identified cosine",
  ],
  [
    "cos([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "cos({0.5})",
    "processed argument",
  ],
  [
    "cos(0.5)",
    "real cosine",
  ],
  [
    "0.8775825618903728",
    "created real",
  ],
]
`;

exports[`cot returns a expression when given non-constant 1`] = `
[
  [
    "cot(x)",
    "identified cotangent",
  ],
  [
    "cot([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "cot({x})",
    "processed argument",
  ],
  [
    "cot(x)",
    "created cotangent",
  ],
]
`;

exports[`cot returns the appropriate real value when given real 1`] = `
[
  [
    "cot(0.5)",
    "identified cotangent",
  ],
  [
    "cot([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "cot({0.5})",
    "processed argument",
  ],
  [
    "cot(0.5)",
    "real cotangent",
  ],
  [
    "(tan(0.5)^-1)",
    "identified exponentiation",
  ],
  [
    "([tan(0.5)]^-1)",
    "processing left operand",
  ],
  [
    "tan(0.5)",
    "identified tangent",
  ],
  [
    "tan([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "tan({0.5})",
    "processed argument",
  ],
  [
    "tan(0.5)",
    "real tangent",
  ],
  [
    "0.5463024898437905",
    "created real",
  ],
  [
    "({0.5463024898437905}^[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(0.5463024898437905^{-1})",
    "processed right operand",
  ],
  [
    "(0.5463024898437905^-1)",
    "real exponentiation",
  ],
  [
    "1.830487721712452",
    "created real",
  ],
]
`;

exports[`csc returns a expression when given non-constant 1`] = `
[
  [
    "csc(x)",
    "identified cosecant",
  ],
  [
    "csc([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "csc({x})",
    "processed argument",
  ],
  [
    "csc(x)",
    "created cosecant",
  ],
]
`;

exports[`csc returns the appropriate real value when given real 1`] = `
[
  [
    "csc(0.5)",
    "identified cosecant",
  ],
  [
    "csc([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "csc({0.5})",
    "processed argument",
  ],
  [
    "csc(0.5)",
    "real cosecant",
  ],
  [
    "(sin(0.5)^-1)",
    "identified exponentiation",
  ],
  [
    "([sin(0.5)]^-1)",
    "processing left operand",
  ],
  [
    "sin(0.5)",
    "identified sine",
  ],
  [
    "sin([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "sin({0.5})",
    "processed argument",
  ],
  [
    "sin(0.5)",
    "real sine",
  ],
  [
    "0.479425538604203",
    "created real",
  ],
  [
    "({0.479425538604203}^[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(0.479425538604203^{-1})",
    "processed right operand",
  ],
  [
    "(0.479425538604203^-1)",
    "real exponentiation",
  ],
  [
    "2.085829642933488",
    "created real",
  ],
]
`;

exports[`sec returns a expression when given non-constant 1`] = `
[
  [
    "sec(x)",
    "identified secant",
  ],
  [
    "sec([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "sec({x})",
    "processed argument",
  ],
  [
    "sec(x)",
    "created secant",
  ],
]
`;

exports[`sec returns the appropriate real value when given real 1`] = `
[
  [
    "sec(0.5)",
    "identified secant",
  ],
  [
    "sec([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "sec({0.5})",
    "processed argument",
  ],
  [
    "sec(0.5)",
    "real secant",
  ],
  [
    "(cos(0.5)^-1)",
    "identified exponentiation",
  ],
  [
    "([cos(0.5)]^-1)",
    "processing left operand",
  ],
  [
    "cos(0.5)",
    "identified cosine",
  ],
  [
    "cos([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "cos({0.5})",
    "processed argument",
  ],
  [
    "cos(0.5)",
    "real cosine",
  ],
  [
    "0.8775825618903728",
    "created real",
  ],
  [
    "({0.8775825618903728}^[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(0.8775825618903728^{-1})",
    "processed right operand",
  ],
  [
    "(0.8775825618903728^-1)",
    "real exponentiation",
  ],
  [
    "1.139493927324549",
    "created real",
  ],
]
`;

exports[`sin returns a expression when given non-constant 1`] = `
[
  [
    "sin(x)",
    "identified sine",
  ],
  [
    "sin([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "sin({x})",
    "processed argument",
  ],
  [
    "sin(x)",
    "created sine",
  ],
]
`;

exports[`sin returns the appropriate real value when given real 1`] = `
[
  [
    "sin(0.5)",
    "identified sine",
  ],
  [
    "sin([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "sin({0.5})",
    "processed argument",
  ],
  [
    "sin(0.5)",
    "real sine",
  ],
  [
    "0.479425538604203",
    "created real",
  ],
]
`;

exports[`tan returns a expression when given non-constant 1`] = `
[
  [
    "tan(x)",
    "identified tangent",
  ],
  [
    "tan([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "tan({x})",
    "processed argument",
  ],
  [
    "tan(x)",
    "created tangent",
  ],
]
`;

exports[`tan returns the appropriate real value when given real 1`] = `
[
  [
    "tan(0.5)",
    "identified tangent",
  ],
  [
    "tan([0.5])",
    "processing argument",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "tan({0.5})",
    "processed argument",
  ],
  [
    "tan(0.5)",
    "real tangent",
  ],
  [
    "0.5463024898437905",
    "created real",
  ],
]
`;
